### PR TITLE
Fix infra clone handle search tables

### LINKED ISSUE #6716: No switch after France infrastructure duplication
### What happened?

The user duplicated the France infrastructure several times. Most of the times, the duplication went well and the duplicated infrastructure seemed to have all the objects of the initial one. However, on one duplication, I could not see any switch.
I did not see any error message.

### What did you expect to happen?

I expected the duplication to be fully completed with all the objects.

### How can we reproduce it (as minimally and precisely as possible)?

Hard to reproduce...

### What operating system, browser and environment are you using?

- **Browser**: Firefox
- **OS**:  Windows ?
- **Env**: Integration

### OSRD version (top right corner `Account` button > `Informations`)

dfc9e4c

### LINKED ISSUE #7663: When cloning an infra, the BP/CC are not displayed anymore on the map
### What happened?

I clone France infra (which has neutral sections). I can not see neutral sections on the cloned infra.

### What did you expect to happen?

I should be able to see the neutral sections if they have been correctly cloned.
If they have not, then they should be correctly cloned.

### How can we reproduce it (as minimally and precisely as possible)?

1. Go to the reference map
2. Select infra france locked and display neutral sections (you should see them)
3. Clone infra france
4. The neutral sections are not displayed

### What operating system, browser and environment are you using?

- **Browser**: Firefox v109
- **OS**: Nixos v22.11
- **Env**: Local

### OSRD version (top right corner `Account` button > `Informations`)

.

### PR DESCRIPTION
**Pull request by flomonster**
# What is done

- The clone operation is done in an sql transaction (close #6716)
- We disable search triggers and fill the tables manually
  - Fix: Can't search in a cloned infra
- Clone the `layer_neutral_section` (fix #7663)
- Clone the `layer_neutral_sign`
- Clone the `layer_psl_sign`

> [!IMPORTANT]
> We have a minor drawback. While an infra is cloned we take a full lock on track, signal and op tables.
> This temporarily blocks several types of operation.

# Profiling

> [!NOTE]
> Time wasted duplicating forgotten layers is regained by deactivating triggers.

### Before

**TOTAL**: 19s

- Clone `infra_object_track_section` in 2002ms
- Clone `infra_layer_track_section` in 900ms
- Clone `infra_object_signal` in 2490ms
- Clone `infra_layer_signal` in 409ms
- Clone `infra_object_speed_section` in 810ms
- Clone `infra_layer_speed_section` in 6319ms
- Clone `infra_object_detector` in 474ms
- Clone `infra_layer_detector` in 566ms
- Clone `infra_object_neutral_section` in 14ms
- Clone `infra_object_switch` in 260ms
- Clone `infra_layer_switch` in 321ms
- Clone `infra_object_extended_switch_type` in 0ms
- Clone `infra_object_buffer_stop` in 22ms
- Clone `infra_layer_buffer_stop` in 26ms
- Clone `infra_object_route` in 545ms
- Clone `infra_object_operational_point` in 896ms
- Clone `infra_layer_operational_point` in 281ms
- Clone `infra_object_electrification` in 24ms
- Clone `infra_layer_electrification` in 402ms
- Clone `errors` in 2753ms

### After

**TOTAL**: 19s

- Clone `infra_object_track_section` in 786ms 🚀
- Clone `infra_layer_track_section` in 877ms
- Clone `infra_object_signal` in 329ms 🚀
- Clone `infra_layer_signal` in 418ms
- Clone `infra_object_speed_section` in 1054ms
- Clone `infra_layer_speed_section` in 6596ms ⚠️ (We should try to reduce the number of speed sections)
- Clone `infra_object_detector` in 444ms
- Clone `infra_layer_detector` in 560ms
- Clone `infra_object_neutral_section` in 14ms
- Clone `infra_object_switch` in 230ms
- Clone `infra_layer_switch` in 303ms
- Clone `infra_object_extended_switch_type` in 0ms
- Clone `infra_object_buffer_stop` in 22ms
- Clone `infra_layer_buffer_stop` in 26ms
- Clone `infra_object_route` in 512ms
- Clone `infra_object_operational_point` in 90ms
- Clone `infra_layer_operational_point` in 276ms 🚀
- Clone `infra_object_electrification` in 24ms
- Clone `infra_layer_electrification` in 378ms
- Clone `errors` in 2647ms

New:
- Clone `search_signal` in 1707ms 🐢
- Clone `search_track` in 10ms
- Clone `search_operational_point` in 314ms 🐢
- Clone `infra_layer_neutral_section` in 15ms
- Clone `infra_layer_psl_sign` in 174ms
- Clone `infra_layer_neutral_sign` in 52ms

### CODE REVIEW

File: editoast/src/modelsv2/infra.rs [hunk 1]
```diff
@@ -235,11 +262,6 @@ impl Infra {
 
         conn.build_transaction()
             .run(|conn| Box::pin(async {
-                // Lock the table to avoid infra edition while the trigger is disabled
-                sql_query("LOCK TABLE infra_object_track_section IN SHARE ROW EXCLUSIVE MODE")
-                    .execute(conn)
-                    .await?;
```
Comment (Castavo) [inline editoast/src/modelsv2/infra.rs:241]: Is this removed because it's redundant with the transaction ?

File: editoast/src/modelsv2/infra.rs [hunk 2]
```diff
@@ -125,65 +122,95 @@ impl Infra {
...
+                sql_query(format!(
+                    "INSERT INTO {model_table}(obj_id,data,infra_id) SELECT obj_id,data,$1 FROM {model_table} WHERE infra_id = $2"
+                ))
+                .bind::<BigInt, _>(cloned_infra.id)
+                .bind::<BigInt, _>(self.id)
+                .execute(conn).await?;
+
+                if let Some(layer_table) = get_geometry_layer_table(&object) {
+                    let layer_table = layer_table.to_string();
+                    let sql = if layer_table != get_geometry_layer_table(&ObjectType::Signal).unwrap() {
```
Comment (Castavo) [inline editoast/src/modelsv2/infra.rs]: Why not 
```suggestion
                    let sql = if object != &ObjectType::Signal {
```

**Review by Castavo — COMMENTED** [review]
Neat, ran it on my machine, works great
Why did you remove `try_join_all` ? Is it impossible to do with the transaction ?

Also, it would be even neater if the sql queries to clone the search layers could be generated by the `Search` derive (like an automatic `get_cloning_query()` impl, but that's maybe overkill

File: editoast/src/modelsv2/infra.rs (hunk 1 above)
Comment (flomonster) [reply editoast/src/modelsv2/infra.rs:241]: It's automatically done by altering the table inside the transaction.

**Review by Castavo — APPROVED** [review]
LGTM

Why did you drop `try_join_all` ? Is it incompatible with the transaction ?

File: editoast/src/modelsv2/infra.rs [hunk 3]
```diff
@@ -125,65 +122,92 @@ impl Infra {
...
+            sql_query("ALTER TABLE infra_object_operational_point ENABLE TRIGGER search_operational_point__ins_trig").execute(conn).await?;
+
+            // Fill search tables
+            sql_query("INSERT INTO search_signal(id, label, line_name, infra_id, obj_id, signaling_systems, settings, line_code) 
+                        SELECT signal.id, label, line_name, $1, search_signal.obj_id, signaling_systems, settings, line_code FROM search_signal 
+                        JOIN infra_object_signal AS signal ON search_signal.obj_id = signal.obj_id and signal.infra_id = $1
+                        WHERE search_signal.infra_id = $2")
+                .bind::<BigInt, _>(cloned_infra.id)
+                .bind::<BigInt, _>(self.id)
+                .execute(conn).await?;
```
Comment (leovalais) [inline editoast/src/modelsv2/infra.rs:180]: Could you add a comment in `search::objects` that these queries will require some modification should any changes be made to their table schema?

**Review by leovalais — APPROVED** [review]
Tested, thanks for the fix!

**Comment by leovalais** [comment]
> LGTM
> 
> Why did you drop `try_join_all` ? Is it incompatible with the transaction ?

A transaction can only apply to a single connection. Besides, the new testing utils makes it harder to test functions that open multiple connections. And an endpoint opening multiple connections can starve others.

I guess it's one of these reasons ¯\_(ツ)_/¯

**Comment by flomonster** [comment]
> > LGTM
> > Why did you drop `try_join_all` ? Is it incompatible with the transaction ?
> 
> A transaction can only apply to a single connection. Besides, the new testing utils makes it harder to test functions that open multiple connections. And an endpoint opening multiple connections can starve others.
> 
> I guess it's one of these reasons ¯_(ツ)_/¯

In addition, there was a comment with the `try_join_all`:

> When creating a connection for each objet, it will a panic with 'Cannot access shared transaction state' in the database pool
> Just one connection fixes it, but partially* defeats the purpose of joining all the requests at the end
>  * AsyncPgConnection supports pipeling within one connection, but it won’t run parallel

File: editoast/src/modelsv2/infra.rs (hunk 3 above)
Comment (flomonster) [reply editoast/src/modelsv2/infra.rs:180]: Done

**Review by SharglutDev — APPROVED** [review]
Lgtm

**Pull request merged**